New nmea0183 parser

fabdrol opened the pull request:
This PR updates `provider/nmea0183-signalk.js` to the new 2.0.2 release of the parser.

fabdrol commented:
@tkurki please review, and merge asap if you see nothing. The new parser has been released already!

fabdrol commented:
The Travis check has failed because the PR was submitted before `2.0.2` of parser had finished uploading to NPM. Any other way of triggering that check again without submitting a new commit?

sbender9 commented:
@fabdrol I restarted the build...

sbender9 commented:
Same issue. I get the same locally. You can go to the build "Details" and click "Restart Build" to try again when it's ready.

fabdrol commented:
Hmm, apparently 2.0.2 didn’t get published; I’ll double check and try that again

fabdrol commented:
Fixed.

tkurki commented on providers/nmea0183-signalk.js:
```diff
@@ -14,81 +14,56 @@
  * limitations under the License.
  */
 
-/* Usage: this is the pipeElement that transforms NMEA0183 input to Signal K deltas
```
Why delete the documentation?

fabdrol replied:
I set out to rewrite it, and then forgot about it. I'll add it back in.

tkurki commented:
The parser is not emitting `nmea0183` events, needed for nmea0183 output over tcp and relaying to serial ports.

https://github.com/SignalK/signalk-server-node/blob/40cb6a14802f7ce4c3cd3a58f7217f6d0b5417cf/providers/nmea0183-signalk.js#L28-L36

fabdrol commented:
@tkurki `nmea0183` event was overlooked, fixed in https://github.com/SignalK/signalk-parser-nmea0183/pull/95

tkurki commented:
@fabdrol are you following this through? Two open pull requests at https://github.com/SignalK/signalk-parser-nmea0183/pulls

fabdrol commented:
@tkurki I added the example and updated the configuration files, plus some small optimisations to the provider. I'm in favour of merging, you concur?

tkurki commented:
You need to reformat with `prettier-standard` per #311 and please squash the commit history, but other than that no objections.

tkurki commented:
Doing it.

tkurki commented:
Formatted & squashed.

tkurki commented:
One minor thing caught my eye: why are we ignoring errors in `this.parser.parse(`${chunk.trim()}\n`).catch(() => {})`?

tkurki commented:
@joabakk @fabdrol I am worried. Literally the first sentence that I happen to look into in more detail with the new parser does not work for my data: https://github.com/SignalK/signalk-parser-nmea0183/pull/98. 

It appears you have changed not just the parser infrastructure but also the logic for individual sentences. In this case you have added cross field checks so that if there is no data in a certain field the whole sentence is silently ignored, even if there would be valid data in another field.

Do you have any idea how many changes of this type you've done in the rewrite?

Is there an easy way to get log messages for sentences that are ignored?

joabakk commented:
I see now that there are numerous ways to check for empty sentences in the hooks, and this specific one seems wrong. We should stick to something like a standard. I like the looks of this one: https://github.com/SignalK/signalk-parser-nmea0183/blob/master/hooks/DSC.js#L71-L80. It is probably easier to copy for new hooks.
Apart from the empty sentence tests, we have not changed the logic of the sentences to my knowledge. The one exception being ROT, which used to have hard coded engine

tkurki commented:
There are other problems:
- gnss@GGA uses incorrect paths
- position@GGA should be object valued

There does not seems to be any schema validation code in the tests 😢 😢 😢
